File: 8. String to Integer (atoi).py
class Solution:
    def myAtoi(self, s: str) -> int:
        s = s.lstrip()  # remove leading whitespaces
        if not s: return 0  # return 0 if string is empty
        if s[0] == '-':
            sign = -1
            s = s[1:]
        elif s[0] == '+':
            sign = 1
            s = s[1:]
        else:
            sign = 1
        result = 0
        for ch in s:
            if not ch.isdigit():
                break
            result = result * 10 + int(ch)
            if result > 2147483647:
                return 2147483647 if sign == 1 else -2147483648
        return sign * result

File: 8. String to Integer (atoi).cpp
class Solution {
public:
    int myAtoi(string s) {
        int i = 0, sign = 1, result = 0;
        while (i < s.size() && s[i] == ' ') i++;
        if (i < s.size() && (s[i] == '-' || s[i] == '+')) {
            sign = (s[i++] == '-') ? -1 : 1;
        }
        while (i < s.size() && isdigit(s[i])) {
            if (result > INT_MAX / 10 || (result == INT_MAX / 10 && s[i] - '0' > INT_MAX % 10)) {
                return (sign == 1) ? INT_MAX : INT_MIN;
            }
            result = result * 10 + (s[i++] - '0');
        }
        return result * sign;
    }
};
